### PR TITLE
Feature/custom logging module

### PR DESCRIPTION
**Pull request by natefleming**
This PR adds supports for rolling logs and periodic uploading to Databricks volumes based on file size or duration.

This implementation uses a custom logging.handler.Handler implementation as well as the "structlog" library

* By default logs are rolled after 1 hour OR 10MB and 5 local copies are kept. All rolled over logs are uploaded 
* initialize_logging(...) - Should be called once to initialize 
* Files are uploaded with the jsonl extension
* Integrates StructLog with Standard Logging allowing external log messages to be captured
* get_logger(...) - This factory will inspect the call stack to determine a log name if not provided

StructLog allows for arbitrary key value pairs to be provided in log messages. These will all result in table columns

Example Usage with Defaults (rollover after 5 min OR 10 MB file size)

```python

from mlflow_extensions.log import initialize_logging, LogLevel, LogConfig

log_config: LogConfig = LogConfig(level=LogLevel.DEBUG, filename="./logs/scratchpad.log", archive_path="/Volumes/catalog/default/logs", when="m", interval=2)
initialize_logging(log_config)


logger: Logger = get_logger()
logger.debug("This is the message", foo="bar", baz="qux")
logger.info("This is the message", foo="bar", baz="qux")
logger.warning("This is the message", foo="bar", baz="qux")
logger.critical("This is the message", foo="bar", baz="qux")
logger.critical("This is the message", foo="bar", baz="qux")
logger.critical("This is the message", foo="bar", baz="qux")
```

Signature for Handler Factory:

```python
def rotating_volume_handler(
    filename: str,
    volume_path: str,
    databricks_host: Optional[str] = None,
    databricks_token: Optional[str] = None,
    when: str = "h",
    interval: int = 1,
    max_bytes: int = DEFAULT_MAX_BYTES,
    backup_count: int = DEFAULT_BACKUP_COUNT,
    encoding: Optional[str] = None,
    delay: bool = False,
    utc: bool = False,
    at_time=None,
    errors=None,
) -> SizeAndTimedRotatingVolumeHandler:
...
```

### CODE REVIEW

**Comment by natefleming** [comment]
Example Message:

```json
{"message": "This is the message", "taskName": null, "foo": "bar", "baz": "qux", "level": "debug", "logger": "__main__", "host": "W7633NGQKK", "ip": "143.244.220.150", "version": "0.10.1.dev9+gb58224f.d20240910", "timestamp": "2024-09-11T15:05:19.032372Z"}
```

**Review by stikkireddy — CHANGES_REQUESTED** [review]

**Comment by stikkireddy** [comment]
> Should the catalog, schema and volume be created if they don't exist?

no i dont think we should create anything, or delete anything on behalf of the user like schema or catalog. we should just error gracefully during deployment. and ask the user to consciously do it.

File: mlflow_extensions/log/handlers.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+class VolumeRotator:
+
+    def __init__(
+        self,
+        volume_path: str,
+        databricks_host: Optional[str],
+        databricks_token: Optional[str],
+    ) -> None:
+        from mlflow.utils.databricks_utils import get_databricks_host_creds
```
[the remark this answers is not in the document]
Comment (natefleming) [reply mlflow_extensions/log/handlers.py]: Thanks for the suggestions.

I put an import guard around the mlflow stuff (I think that function is deprecated anyway) and it prefers env variables.

I also added configuration of the volume through env vars.

The environment variables use to determine volume path are also logged as fields the event messages

Does something like this work?


```python
def _get_volume_path(volume_path: Optional[str]) -> Optional[str]:
    volume_path = volume_path or os.environ.get("LOGGING_VOLUME_PATH")
    if volume_path is None:
        volume: str = os.environ.get("LOGGING_VOLUME")
        volume_path: str = _full_volume_name_to_path(volume)
        model_name: str = os.environ.get("LOGGING_MODEL_NAME")
        endpoint_id: str = os.environ.get("LOGGING_ENDPOINT_ID")
        run_id: str = os.environ.get("LOGGING_RUN_ID")

        if volume_path is not None:
            volume_path = "/".join(
                [
                    x
                    for x in [volume_path, model_name, endpoint_id, run_id]
                    if x is not None
                ]
            )

    return volume_path
```

```python
def _get_databricks_host_creds(
    databricks_host: Optional[str], databricks_token: Optional[str]
) -> Tuple[str, str]:
    if databricks_host is None or databricks_token is None:
        try:
            from mlflow.utils.databricks_utils import get_databricks_host_creds

            databricks_host = os.environ.get(
                "DATABRICKS_HOST", get_databricks_host_creds().host
            )
            databricks_token = os.environ.get(
                "DATABRICKS_TOKEN", get_databricks_host_creds().token
            )
        except ImportError:
            databricks_host = os.environ.get("DATABRICKS_HOST")
            databricks_token = os.environ.get("DATABRICKS_TOKEN")

    return databricks_host, databricks_token
```

**Comment by stikkireddy** [comment]
i will try to close this and the other PR out by tomorrow. This log rotation will be helpful also for lakehouse apps :D. Ideally we can reuse most of it as a one off file.

**Comment by natefleming** [comment]
Sounds good. I hacked on it a bit today. I'm not quite sure where the best place to integrate it. I also think there might be a bug in filtering std logs but it should be close

**Review by stikkireddy — APPROVED** [review]
LGTM

**Pull request merged**